src/lighthouse.rs

```diff
@@ -138,16 +141,35 @@ fn quorum_compute(
     // Sort by replica ID to get a consistent ordering across runs.
     candidate_participants.sort_by_key(|p| p.replica_id.clone());
 
+    let shrink_only = healthy_participants
+        .iter()
+        .any(|(_, details)| details.member.shrink_only);
+
     let metadata = format!(
-        "[{}/{} participants healthy]",
+        "[{}/{} participants healthy][shrink_only={}]",
         healthy_participants.len(),
-        state.participants.len()
+        state.participants.len(),
+        shrink_only,
     );
 
     // Check if we can use the previous quorum.
+    // TODO: do we still need this given we have heartbeats?
     if state.prev_quorum.is_some() {
         let prev_quorum = state.prev_quorum.as_ref().unwrap();
 
+        let prev_replica_ids: HashSet<&String> = prev_quorum
+            .participants
+            .iter()
+            .map(|p| &p.replica_id)
+            .collect();
+
+        if shrink_only {
+            candidate_participants = candidate_participants
+                .into_iter()
+                .filter(|p| prev_replica_ids.contains(&p.replica_id))
+                .collect();
+        }
+
         // Fast quorum is when all previous participants are still in the quorum
         // and we have enough participants to form a quorum.
         let is_fast_quorum = prev_quorum
```

Comment:
instead of prev_quorum, should this check `candidate_participants` are all in prev_replica_ids?

Reply:
In fast quorum we do want to allow it to grow which is why we don't check that all candidate_participants are in prev_replica_ids. There could be new candidates

I may also just delete this given the changes to join timeout below